pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyServiceStarter.java

```diff
@@ -240,7 +240,10 @@ public static void addWebServerHandlers(WebServer server,
                                      ProxyConfiguration config,
                                      ProxyService service,
                                      BrokerDiscoveryProvider discoveryProvider) throws Exception {
-        server.addServlet("/metrics", new ServletHolder(MetricsServlet.class),
+        PrometheusMetricsServlet metricsServlet =
+                new PrometheusMetricsServlet(-1L, config.getClusterName());
+        service.getMetricsProviders().forEach(metricsServlet::addRawMetricsProvider);
```

Comment:
Is there a race here that could prevent some metrics providers from getting added to the `metricsServlet`? In looking at the broker's implementation, it ensures that the `metricsServlet` is updated if the servlet has already been created.

https://github.com/apache/pulsar/blob/b0213b225f194b47a5dcd63ef4a26f55c4c820b6/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java#L1518-L1527

Reply:
I've added support for late registration similar to what is done in the broker. Can you check again ?

Reply:
Thanks @cbornet. That looks good to me.